src/lender/tranche/operator/proportional.sol

```diff
@@ -0,0 +1,173 @@
+    mapping (address => uint) public currentSupplyLimit;
+
+    // expressed in totalCurrencyReturned notation
+    mapping (address => uint) public currencyRedeemed;
+
+    // expressed in totalPrincipalReturned notation
+    mapping (address => uint) public principalRedeemed;
+
+    bool public supplyAllowed  = true;
+
+    uint public totalCurrencyReturned;
+    uint public totalPrincipalReturned;
+
+    // total invested amount of currency and total supply of token
+    // contract implements a fixed supply token price of ONE
+    uint public totalTrancheVolume;
+
+    constructor(address tranche_, address assessor_, address distributor_) public {
+        wards[msg.sender] = 1;
+        tranche = TrancheLike(tranche_);
+        assessor = AssessorLike(assessor_);
+        distributor = DistributorLike(distributor_);
+    }
+
+    /// sets the dependency to another contract
+    function depend(bytes32 contractName, address addr) public auth {
+        if (contractName == "tranche") { tranche = TrancheLike(addr); }
+        else if (contractName == "assessor") { assessor = AssessorLike(addr); }
+        else if (contractName == "distributor") { distributor = DistributorLike(addr); }
+        else revert();
+    }
+
+    function file(bytes32 what, bool supplyAllowed_) public auth {
+        if(what == "supplyAllowed") {
+            supplyAllowed = supplyAllowed_;
+        }
+    }
+
+    function updateReturned(uint currencyReturned, uint principalReturned) public auth {
+        totalCurrencyReturned += currencyReturned;
+        totalPrincipalReturned += principalReturned;
+    }
+
+    /// defines the max amount of currency for supply
+    function approve(address usr, uint currencyAmount) external auth {
+        supplyMaximum[usr] = currencyAmount;
+        currentSupplyLimit[usr] = currencyAmount;
+    }
+
+    /// only approved investors can supply and approved
+    function supply(uint currencyAmount) external note {
+        require(supplyAllowed);
+
+        require(currentSupplyLimit[msg.sender] >= currencyAmount, "not-enough-currency");
+        currentSupplyLimit[msg.sender] = safeSub(currentSupplyLimit[msg.sender], currencyAmount);
+
+        require(assessor.supplyApprove(address(tranche), currencyAmount), "supply-not-approved");
+
+        // pre-defined tokenPrice of ONE
+        tranche.supply(msg.sender, currencyAmount, rdiv(currencyAmount, ONE));
+
+        // todo we don't need the variable if first loan starts after all investors supplied
+        // instead tranche.balance could be used
+        totalTrancheVolume = safeAdd(totalTrancheVolume, currencyAmount);
+
+        distributor.balance();
+    }
+
+    /// redeem is proportional allowed
+    function redeem(uint tokenAmount) external note {
+        distributor.balance();
+        (uint currencyAmount, uint currencyRedeemed_,uint  principalRedeemed_) = calcRedeemCurrencyAmount(msg.sender, tokenAmount);
+        currencyRedeemed[msg.sender] = currencyRedeemed_;
```

Comment:
not better changing the state after calling tranche.redeem & receiving approval from assessor?